contracts/RootChain/RootChain.sol

```diff
@@ -187,20 +197,41 @@ contract RootChain {
             require(merkleHash.checkMembership(txPos[1], childChain[txPos[0]].root, proof));
         }
 
-        // one-to-one mapping between priority and exit
-        require(exits[priority].owner == address(0));
-        require(exits[priority].amount == 0);
+        // check that the UTXO's two direct inputs have not been previously exited
+        validateExitInputs(txList);
 
+        // create new started/pending exit after passing all previous checks
         exitsQueue.insert(priority);
-
         exits[priority] = exit({
-            owner: txList[10 + 2 * txPos[2]].toAddress(),
             amount: txList[11 + 2 * txPos[2]].toUint(),
             utxoPos: txPos,
-            created_at: block.timestamp
+            created_at: block.timestamp,
+            owner: txList[10 + 2 * txPos[2]].toAddress(),
+            state: 1
         });
     }
 
+    /// For any attempted exit of an UTXO, validate that the UTXO's two inputs have not
+    /// been previously exited. If UTXO's inputs are in the exit queue, those inputs'
+    /// exits are deleted from the exit queue and the current UTXO's exit remains valid.
+    function validateExitInputs(RLP.RLPItem[] memory txList)
+        private
+    {
+        for (uint256 i = 0; i < 2; i++) {
+            uint256 txInputBlkNum = txList[5 * i + 0].toUint();
+            uint256 txInputIndex = txList[5 * i + 1].toUint();
+            uint256 txInputOutIndex = txList[5 * i + 2].toUint();
+            uint256 txInputPriority = 1000000000 * txInputBlkNum + 10000 * txInputIndex + txInputOutIndex;
+
+            // check that the UTXO's inputs have not been finalized
+            require(exits[txInputPriority].state != 3);
+
+            if (exits[txInputPriority].state == 1 && exits[txInputPriority].owner != address(0)) {
```

Comment:
Why is there a check for a zero-address?

Reply:
Yeah, if `state = 1` then the owner should exist.